CSL-2046 Student details page fixes from Q/A issues

klown opened the pull request:
Q/A issues fixed:
- Reaction panel text, just above the reaction wheel
- Content of the reaction modal dialog:
  - If the reaction's free response is empty, do not show "None",
  - The do-not-enter sign is prepended to titles that the teacher is not authorized to view,
  - If any book is unauthorized, the unauthorized legend is displayed at the bottom of the dialog.
- Added spaces around the "/" in the Learning column that separates the amount learned and the free response,
- Handles corrupt data specifically when the last viewed time is `None`.  In that case, the book(s) is/are listed last.

JIRA:  https://castudl.atlassian.net/browse/CSL-2046

[CSL-2046]: https://castudl.atlassian.net/browse/CSL-2046?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

klown committed:
CSL-2046: Fix the Reactions panel text

klown committed:
CSL-2046: Fix display of student's detail reaction modal dialog

1. If the reaction's free response is empty (`None`, empty string,
   or all white space), nothing is shown in the dialog.  Previously,
   if it was `None`, then the string "None" was shown,
2. For each book that the teacher is not authorized to view, add
   the do-no-enter sign to the front of the book's title,
3. If any book is unauthorized, show the unauthorized legend at
   the bottom of the dialog.

klown committed:
CSL-2046: Fix spacing between amount learned and comprehension free response

In the Reading details panel's "Learning" column, add a space on
either side of the "/" character that divides the amount learned
and the student's free response.

klown committed:
CSL-2046: Fix corrupt data when the last viewed time is `None`

For the Reading details table on the studet details page, put any
books whose `last_view` time is `None` at the bottom of the list,
and indicate that the date is "Unknown".

klown commented:
@bgoldowsky this is ready for review.

The fix for the corrupted `last_view` times is a little brute force and I'll think some more on it.  In summary, I added a loop to stash-and-remove all of the items with null `last_view` times, sort the remaining using the existing lambda, and then append the stashed items to the end of th sorted list.  I did not use `.desc(nulls_last=True)` because I wasn't sure it was relevant anymore.  The technique is sufficient, if a bit ugly.

bgoldowsky commented on src/library/models.py:
```diff
@@ -744,7 +744,14 @@ def get_reading_data(cls, period: Period, days=0, sort='name', books_sort='title
+                            last_view_null.append(abook)
+                            book_list.remove(abook)
+                    book_list.sort(reverse=True, key=lambda item: item['last_view'])
+                    book_list.extend(last_view_null)
```
Sure, that works, or you could do something like `sort(key=lambda item: item['last_view'] or a_very_old_date)`

klown replied:
That's similar to an approach I abandoned.  During the earlier creation of the map, from which the `book_list` is derived, set the `last_view` property to a year ago if the database record's value was null.  Then, the sort would just proceed as before, putting all the items from the fake year ago at the end.  The problem was that you could not distinguish "a year ago" from a legitimate date.  If we had a constant set to `VERY_OLD_DATE`, that could act as a marker.  I'll think some more on that.

bgoldowsky commented on src/roster/templates/roster/partial/student_details_reading_details.html:
```diff
@@ -67,9 +67,21 @@ <h4 class="h2 mb-0">Reading details</h4>
+                                {% else %}
+                                    Unknown
+                                {% endif %}
+                            </td>
```
It might work here to just say `{ book.last_view | naturalday:"D, M d" | default:"Unknown" }}

klown replied:
It does!  That makes it cleaner.

bgoldowsky commented on src/shared/templates/shared/partial/modal_student_details_affect.html:
```diff
@@ -7,13 +7,25 @@
     </ul>
+    {% if any_unauthorized_book %}
+    <p class="form-text"><em>Readings with the symbol <span class="icon-block" title="Not authorized to view"></span> were uploaded by the student.</em></p>
+    {% endif %}
```
This looks great.  We want to also do the same thing on the main page, if this icon appears in the readings table.

klown replied:
Okay.  The simplest way is to show it always, similar to how the "Assigned" legend is shown at the top.  Due to paging, some pages of results will have no assigned books listed, yet "Assigned" is always present.  The "Unauthorized" legend could behave the same way -- always shown.

klown replied:
The latest push addresses your latest three comments, @bgoldowsky.

In addition, I did write a version of the code that supports a conditional presentation of the "Unauthorized" legend at the bottom of the reading details table.  However, I still lean towards it being displayed unconditionally, so that is the version I pushed.  I do have a copy of the conditional version and can swap it in if necessary.

klown replied:
I wrote:

> In addition, I did write a version of the code that supports a conditional presentation of the "Unauthorized" legend at the bottom of the reading details table. However, I still lean towards it being displayed unconditionally, so that is the version I pushed. I do have a copy of the conditional version and can swap it in if necessary.

Rather than store that version locally on my machine, I've made another PR #369 that includes the code for showing the "unauthorized" legend only when an unauthorized book is shown in the table.  PR #369 is based on the branch for this PR.  Thus, the code is the same for the two PRs except for the conditional display of the legend.

bgoldowsky reviewed (commented):
Looking good, I added a couple of comments.

klown committed:
CSL-2046: Modify based on code review
- use `VERY_LONG_TIME_AGO` for null `last_view` times when sorting,
- clean up template conditional for displaying "Unknown" for null
  `last_view` times,
- add "Unauthorized" legend to bottom of reading details table.

klown committed:
CSL-2046: Fix `clusive_student` template variable in reading details table

The readings detail table shows "No reading" for the current time frame
but only if there is a `current_student` template variable set.  It was
not set properly for ajax calls that changed the table's sorting order
(by title, active time, or last read).
